### PR TITLE
refactor webcheck

### PR DESCRIPTION
**Pull request by requilence**
- split runWebCheck()

gocyclo before:
```
28 frontman (*Frontman).runWebCheck webcheck.go:90:1
8 frontman getTextFromHTML webcheck.go:20:1
```

gocyclo after:
```
15 frontman (*Frontman).runWebCheck webcheck.go:120:1
8 frontman getTextFromHTML webcheck.go:46:1
```

### CODE REVIEW

File: webcheck.go [hunk 1]
```diff
@@ -67,9 +68,29 @@ func (fm *Frontman) initHttpTransport() {
...
+		var err error
+		// read and search in raw file
+		text, err = ioutil.ReadAll(reader)
+		if err != nil {
+			return fmt.Errorf("got error while reading response body: %s", err.Error())
+		}
+	}
+
+	if !bytes.Contains(text, []byte(pattern)) {
+		return fmt.Errorf("pattern '%s' not found %s", pattern)
```
[the remark this answers is not in the document]
Comment (requilence) [reply webcheck.go]: fixed

**Comment by soupdiver** [comment]
why have I been assigned @requilence ?

File: webcheck.go [hunk 2]
```diff
@@ -119,91 +129,86 @@ func (fm *Frontman) runWebCheck(data WebCheckData) (m map[string]interface{}, er
 
 	if data.Method == "POST" && data.PostData != "" {
 		req.Body = ioutil.NopCloser(strings.NewReader(data.PostData))
-		req.Body.Close()
+		// close noop closer to bypass lint warnings
+		_ = req.Body.Close()
```
[the remark this answers is not in the document]
Comment (requilence) [reply webcheck.go]: yes, it always return nil. IDE show me the warning if I don't handle the error this way. Not sure if golangci will do the same

File: webcheck.go [hunk 3]
```diff
@@ -119,91 +129,86 @@ func (fm *Frontman) runWebCheck(data WebCheckData) (m map[string]interface{}, er
...
-	if data.ExpectedHTTPStatus > 0 && resp.StatusCode != data.ExpectedHTTPStatus {
-		m[prefix+"totalTimeSpent_s"] = time.Since(startedConnectonAt).Seconds()
-		return m, fmt.Errorf("bad status code. Expected %d, got %d", data.ExpectedHTTPStatus, resp.StatusCode)
-	}
-
-	m[prefix+"success"] = 1
+	defer func() {
+		// don't care about body close error
+		// because it doesn't affects any kind of checks we have
+		_ = resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (requilence) [reply webcheck.go]: IDE show me the warning if I don't handle the error this way. Not sure if golangci will do the same

File: webcheck.go [hunk 4]
```diff
@@ -119,91 +129,86 @@ func (fm *Frontman) runWebCheck(data WebCheckData) (m map[string]interface{}, er
...
-					err = fmt.Errorf("got error while reading response body: %s", ctx.Err().Error())
-				}
-			} else {
-				err = fmt.Errorf("pattern '%s' not found %s", data.ExpectedPattern, where)
-			}
-		}
-
+		err = isBodyReaderMatchesPattern(resp.Body, data.ExpectedPattern, !data.SearchHTMLSource)
 	} else {
+		// we don't need the content itself
```
[the remark this answers is not in the document]
Comment (requilence) [reply webcheck.go]:
```
if data.ExpectedPattern != "" {
...
} else {
// we don't need the content itself
}
```

BTW I will improve the comment

File: webcheck.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (soupdiver) [reply webcheck.go]: > IDE show me the warning if I don't handle the error this way. Not sure if golangci will do the same

gotcha

File: webcheck.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (soupdiver) [reply webcheck.go]: ✔️

File: webcheck.go [hunk 5]
```diff
@@ -67,43 +68,52 @@ func (fm *Frontman) initHttpTransport() {
...
+
+func (fm *Frontman) runWebCheck(data WebCheckData) (map[string]interface{}, error) {
+	prefix := fmt.Sprintf("http.%s.", data.Method)
+	m := make(map[string]interface{})
+	m[prefix+"success"] = 0
+
+	httpClientWithMaxRedirection := fm.newHTTPClientWithCustomMaxRedirectionsLimit(fm.HTTPCheckMaxRedirects)
 	timeout := fm.HTTPCheckTimeout
 
 	if data.Timeout < timeout {
```
[the remark this answers is not in the document]
Comment (requilence) [reply webcheck.go]: here was a mistake. fixed

File: webcheck.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (requilence) [reply webcheck.go]: fixed

File: webcheck.go [hunk 6]
```diff
@@ -119,91 +129,86 @@ func (fm *Frontman) runWebCheck(data WebCheckData) (m map[string]interface{}, er
...
-	if seconds > 0 {
-		m[prefix+"downloadPerformance_bps"] = int64(float64(bytesReceived) / seconds) // Measure download speed since the request sent
+
+	secondsSinceRequestWasSent := time.Since(wroteRequestAt).Seconds()
+	if secondsSinceRequestWasSent > 0 {
+		m[prefix+"downloadPerformance_bps"] = int64(float64(bytesReceived) / secondsSinceRequestWasSent) // Measure download speed since the request sent
 	}
 
-	return
+	if ctx.Err() == context.DeadlineExceeded {
```
[the remark this answers is not in the document]
Comment (requilence) [reply webcheck.go]: removed duplicate check

File: webcheck.go [hunk 7]
```diff
@@ -119,91 +129,86 @@ func (fm *Frontman) runWebCheck(data WebCheckData) (m map[string]interface{}, er
...
+	*http.Client
+	Err error
+}
+
+func (fm *Frontman) newHTTPClientWithCustomMaxRedirectionsLimit(maxRedirects int) *httpClientAndError {
+	if fm.httpTransport == nil {
+		fm.initHttpTransport()
+	}
+
+	client := &httpClientAndError{&http.Client{Transport: fm.httpTransport}, nil}
```
[the remark this answers is not in the document]
Comment (requilence) [reply webcheck.go]: added a comment

File: webcheck.go [hunk 8]
```diff
@@ -67,43 +68,52 @@ func (fm *Frontman) initHttpTransport() {
 	}
 }
 
-func (fm *Frontman) runWebCheck(data WebCheckData) (m map[string]interface{}, err error) {
-	prefix := fmt.Sprintf("http.%s.", data.Method)
-	m = make(map[string]interface{})
-	m[prefix+"success"] = 0
+func isBodyReaderMatchesPattern(reader io.Reader, pattern string, extractTextFromHTML bool) error {
```
[the remark this answers is not in the document]
Comment (requilence) [reply webcheck.go]: renamed to checkBodyReaderMatchesPattern

File: webcheck.go [hunk 9]
```diff
@@ -100,9 +99,11 @@ func (fm *Frontman) runWebCheck(data WebCheckData) (map[string]interface{}, erro
 	httpClientWithMaxRedirection := fm.newHTTPClientWithCustomMaxRedirectionsLimit(fm.HTTPCheckMaxRedirects)
 	timeout := fm.HTTPCheckTimeout
 
-	if data.Timeout < timeout {
-		timeout = fm.HTTPCheckTimeout
+	// set individual timeout in case it is less than in this check
+	if data.Timeout > 0 && data.Timeout < timeout {
+		timeout = data.Timeout
 	}
+	fmt.Printf("timeout: %.5f\n", timeout)
```
[the remark this answers is not in the document]
Comment (requilence) [reply webcheck.go]: it was fixed in the last commit yesterday. Probably you made a review before it and just submitted it

[the remark this answers is not in the document]
Comment (soupdiver) [reply webcheck.go]: ah okay... I think I was somehow looking at an old state

**Pull request merged**